pkg/disttask/framework/taskexecutor/manager_test.go

```diff
@@ -314,26 +317,141 @@ func TestSlotManagerInManager(t *testing.T) {
 
 	// task1 alloc resource success
 	require.Equal(t, 0, m.slotManager.available)
-	require.Equal(t, map[int64]*slotInfo{
-		taskID1: {taskID: int(taskID1), slotCount: 10},
-	}, m.slotManager.executorSlotInfos)
-	ch <- struct{}{}
+	require.Equal(t, []*proto.Task{task1}, m.slotManager.executorSlotInfos)
+	ch <- nil
 
 	// task1 succeed
 	task1.State = proto.TaskStateSucceed
 	mockTaskTable.EXPECT().GetTaskByID(m.ctx, taskID1).Return(task1, nil)
 	mockInternalExecutor.EXPECT().Close()
+	wg.Wait()
+	require.Equal(t, 10, m.slotManager.available)
+	require.Equal(t, 0, len(m.slotManager.executorSlotInfos))
+	require.True(t, ctrl.Satisfied())
+
+	// ******** Test task occupation ********
+	task1.State = proto.TaskStateRunning
+	var (
+		taskID3 = int64(3)
+		task3   = &proto.Task{
+			ID:          taskID3,
+			State:       proto.TaskStateRunning,
+			Concurrency: 1,
+			Priority:    -1,
+			Step:        proto.StepOne,
+			Type:        "type",
+		}
+	)
+	// 1. task1 alloc success
+	mockTaskTable.EXPECT().HasSubtasksInStates(m.ctx, id, taskID1, proto.StepOne,
+		unfinishedSubtaskStates...).
+		Return(true, nil)
+	mockPool.EXPECT().Run(gomock.Any()).DoAndReturn(runFn)
+	mockTaskTable.EXPECT().HasSubtasksInStates(m.ctx, id, taskID2, proto.StepOne,
+		unfinishedSubtaskStates...).
+		Return(true, nil)
+
+	// mock inside onRunnableTask
+	mockInternalExecutor.EXPECT().Init(gomock.Any()).Return(nil)
+	mockTaskTable.EXPECT().GetTaskByID(m.ctx, taskID1).Return(task1, nil)
+	mockTaskTable.EXPECT().HasSubtasksInStates(m.ctx, id, taskID1, proto.StepOne,
+		unfinishedSubtaskStates...).
+		Return(true, nil)
+	// task1 start running
+	mockInternalExecutor.EXPECT().Run(gomock.Any(), task1).DoAndReturn(func(_ context.Context, _ *proto.Task) error {
+		return <-ch
+	})
+
+	m.onRunnableTasks([]*proto.Task{task1, task2})
+	time.Sleep(2 * time.Second)
+	// task1 alloc resource success
+	require.Equal(t, 0, m.slotManager.available)
+	require.Equal(t, []*proto.Task{task1}, m.slotManager.executorSlotInfos)
+	require.True(t, ctrl.Satisfied())
 
+	// 2. task1 is occupied by task3, task1 start to pausing
+	// 3. task3 is waiting for task1 to be released, and task2 can't be allocated
+	mockTaskTable.EXPECT().HasSubtasksInStates(m.ctx, id, taskID3, proto.StepOne,
+		unfinishedSubtaskStates...).
+		Return(true, nil)
+
+	// the priority of task3 is higher than task2, so task3 is in front of task2
+	m.onRunnableTasks([]*proto.Task{task3, task2})
+	time.Sleep(2 * time.Second)
+	require.Equal(t, 0, m.slotManager.available)
+	require.Equal(t, []*proto.Task{task1}, m.slotManager.executorSlotInfos)
+	require.True(t, ctrl.Satisfied())
+
+	mockTaskTable.EXPECT().GetTaskByID(m.ctx, taskID1).Return(task1, nil)
+	mockTaskTable.EXPECT().HasSubtasksInStates(m.ctx, id, taskID1, proto.StepOne,
+		unfinishedSubtaskStates...).
+		Return(true, nil)
+	mockInternalExecutor.EXPECT().Close()
+
+	// 4. task1 is released, task3 alloc success, start to run
+	ch <- context.Canceled
+	time.Sleep(time.Second)
```

Comment:
can't rely on time.sleep to wait free()

Reply:
The `time.Sleep` is wait here

<img width="695" alt="image" src="https://github.com/pingcap/tidb/assets/23556834/2d9371aa-00d6-4635-9bea-46a028e489e1">

Reply:
Time is unreliable, still have chance of flaky test.

Reply:
> Time is unreliable, still have chance of flaky test.

For what you've said, I think it needs to be analyzed on a case-by-case basis. There's obviously a waiting for `checkTime`, and I don't see any problem with waiting. What I can do is that refactor it to `require.Eventually`. What do you think?

Reply:
https://github.com/pingcap/tidb/pull/49713/commits/4b140b90201266c6e8925298e01c454bb55f142d